dm/worker/task_checker.go

```diff
@@ -0,0 +1,283 @@
+// resume the task and retry
+func isRetryableError(err *pb.ProcessError) bool {
+	// not elegant code, because TiDB doesn't expose some error
+	unsupportedDDLMsgs := []string{
+		"can't drop column with index",
+		"unsupported add column",
+		"unsupported modify column",
+		"unsupported modify",
+		"unsupported drop integer primary key",
+	}
+	if err.Type == pb.ErrorType_ExecSQL {
+		for _, msg := range unsupportedDDLMsgs {
+			if strings.Contains(err.Msg, msg) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (tsc *realTaskStatusChecker) getResumeStrategy(taskStatus *pb.SubTaskStatus, duration time.Duration) ResumeStrategy {
+	// task that is not paused or paused manually, just ignore it
+	if taskStatus == nil || taskStatus.Stage != pb.Stage_Paused || taskStatus.Result == nil || taskStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range taskStatus.Result.Errors {
+		if !isRetryableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.latestResume) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	var (
+		// paused task found in this check round, historical paused task that
+		// can't be automatically resumed will not increase the counter
+		abnormal = 0
+		// resumed task counter
+		resumed = 0
+	)
+	tasks := tsc.w.getPausedSubTasks()
+	duration := tsc.bf.Current()
+	// rotate tsc.block and an empty map, so we can remove resumed block task
+	block := make(map[string]time.Time)
+	for k, v := range tsc.block {
+		block[k] = v
+	}
+	tsc.block = make(map[string]time.Time)
+	for _, task := range tasks {
+		strategy := tsc.getResumeStrategy(task, duration)
+		if strategy == ResumeIgnore {
+			continue
+		}
+		abnormal++
+		if strategy == ResumeNoSense {
+			if ts, ok := block[task.Name]; ok {
+				abnormal--
+				tsc.block[task.Name] = ts
+				tsc.l.Warn("task can't auto resume", zap.String("task", task.Name), zap.Duration("paused duration", time.Since(ts)))
+			} else {
+				tsc.block[task.Name] = time.Now()
+				tsc.l.Warn("task can't auto resume", zap.String("task", task.Name))
+			}
+			continue
+		}
+		if strategy == ResumeSkip {
+			tsc.l.Warn("skip auto resume task", zap.String("task", task.Name), zap.Time("latestResume", tsc.latestResume), zap.Reflect("duration", duration))
+			continue
+		}
+		opLogID, err := tsc.w.operateSubTask(&pb.TaskMeta{
+			Name: task.Name,
+			Op:   pb.TaskOp_AutoResume,
+		})
+		if err != nil {
+			tsc.l.Error("dispatch auto resume task failed", zap.String("task", task.Name), zap.Error(err))
+		} else {
+			resumed++
+			tsc.latestResume = time.Now()
+			tsc.l.Info("dispatch auto resume task", zap.String("task", task.Name), zap.Int64("opLogID", opLogID))
+		}
+	}
+	if abnormal > 0 {
+		tsc.latestNormalTime = time.Now()
```

Comment:
Is `if abnormal > 0` conflict with `the latest time that checker observes no abnormal task`?